src/reformat.c

```diff
@@ -109,50 +109,60 @@ avifBool avifPrepareReformatState(const avifImage * image, const avifRGBImage *
     state->rgbMaxChannel = (1 << rgb->depth) - 1;
     state->yuvMaxChannelF = (float)state->yuvMaxChannel;
     state->rgbMaxChannelF = (float)state->rgbMaxChannel;
-    state->uvBias = 1 << (image->depth - 1);
+    state->yBias = state->yuvRange == AVIF_RANGE_LIMITED ? (float)(16 << (state->yuvDepth - 8)) : 0.f;
+    state->uvBias = (float)(1 << (state->yuvDepth - 1));
+    state->yRange = state->yuvRange == AVIF_RANGE_LIMITED ? (float)(219 << (state->yuvDepth - 8)) : state->yuvMaxChannelF;
+    state->uvRange = state->yuvRange == AVIF_RANGE_LIMITED ? (float)(224 << (state->yuvDepth - 8)) : state->yuvMaxChannelF;
 
     uint32_t cpCount = 1 << image->depth;
     for (uint32_t cp = 0; cp < cpCount; ++cp) {
-        int unormY = cp;
-        int unormUV = cp;
-        if (image->yuvRange == AVIF_RANGE_LIMITED) {
-            unormY = avifLimitedToFullY(image->depth, unormY);
-            unormUV = avifLimitedToFullUV(image->depth, unormUV);
-        }
-        state->unormFloatTableY[cp] = (float)unormY / state->yuvMaxChannelF;
-        if (state->mode == AVIF_REFORMAT_MODE_IDENTITY) {
-            state->unormFloatTableUV[cp] = (float)unormY / state->yuvMaxChannelF;
-        } else {
-            state->unormFloatTableUV[cp] = ((float)unormUV - state->uvBias) / state->yuvMaxChannelF;
+        int unorm = cp;
+        switch (state->mode) {
```

Comment:
Nit: Please change the switch statement back to an if statement (as in the original code). I believe the if statement will be compiled to more efficient code in this case.

In fact, we should put the if statement outside and put the for loop inside the "if" and "else" branches, so that the for loop bodies have no conditionals.

Note: It is instructive to look at the assembly code that the compiler generates for this kind of statement. Although the switch statement exhaustively lists all possible values of the enum type, the compiler will still generate code equivalent to the following:

    if ((state->mode == AVIF_REFORMAT_MODE_YCGCO) || (state->mode == AVIF_REFORMAT_MODE_YUV_COEFFICIENTS)) {
        state->unormFloatTableY[cp] = ((float)unorm - state->yBias) / state->yRange;
        state->unormFloatTableUV[cp] = ((float)unorm - state->uvBias) / state->uvRange;
    } else if (state->mode == AVIF_REFORMAT_MODE_IDENTITY) {
        state->unormFloatTableY[cp] = ((float)unorm - state->yBias) / state->yRange;
        state->unormFloatTableUV[cp] = ((float)unorm - state->yBias) / state->yRange;
    }

I.e., the compiler will transform the last case in the switch statement to an "else if" rather than an "else".

Reply:
I see. By the way I've done an experimental implementation to support all matrix and that needs 6 reformat modes. In that case, should we switch to a switch statement, or keep using if-else chain and sort the most common case first?

Reply:
I did some tests on this: I set up 6 branches with exponential distributed probability, in this case jumptable based switch and if chain ordered by probability resulted in similar speed. Specially Clang is generating jumptable even for if chain.
Based on the result, I may use switch statement when there's more modes added. Using switch also enables compiler to check if the reformat route is correctly implemented when new matrix coefficients are added in the future.

Reply:
Thank you for the report of your experiments!

I suggested if-else statements in this PR because they are simpler than the equivalent switch statements to handle two options. As soon as we need to handle three or more options, then switch statements are often clearer.

I should correct/clarify my previous comment "I believe the if statement will be compiled to more efficient code in this case."

For a switch statement that lists all values of an enum type exhaustively but does not have a default case, the compiler still treats it as if the switch statement had a default case that does nothing. This is why gcc warns that `unorm` may be used uninitialized in the following version of `avifReformatStateUVToUNorm()`, even though the switch statement assigns a value to `unorm` in all possible cases:

```
static int avifReformatStateUVToUNorm(avifReformatState * state, float v)
{
    int unorm;

    // ...

    switch (state->mode) {
        case AVIF_REFORMAT_MODE_IDENTITY:
            unorm = (int)avifRoundf(v * state->rangeY + state->biasY);
            break;
        case AVIF_REFORMAT_MODE_YUV_COEFFICIENTS:
        case AVIF_REFORMAT_MODE_YCGCO:
            unorm = (int)avifRoundf(v * state->rangeUV + state->biasUV);
            break;
    }

    return AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
}
```

Similarly, gcc warns that control reaches end of non-void function in the following version of version of `avifReformatStateUVToUNorm()`:

```
static int avifReformatStateUVToUNorm(avifReformatState * state, float v)
{
    int unorm;

    // ...

    switch (state->mode) {
        case AVIF_REFORMAT_MODE_IDENTITY:
            unorm = (int)avifRoundf(v * state->rangeY + state->biasY);
            return AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
        case AVIF_REFORMAT_MODE_YUV_COEFFICIENTS:
        case AVIF_REFORMAT_MODE_YCGCO:
            unorm = (int)avifRoundf(v * state->rangeUV + state->biasUV);
            return AVIF_CLAMP(unorm, 0, state->yuvMaxChannel);
    }
}
```

Note: In my experiments, clang does not warn about these two versions of avifReformatStateUVToUNorm(). Only gcc does. (I have gcc 10.2.0.) That may be a sign that clang and gcc handle this kind of switch statement differently.

By using a default case in the switch statement (or equivalently, ending an if-else-if chain with an "else"), we avoid this issue. But the benefits of this kind of switch statement may be more important than the minor annoyance of dealing with these compiler warnings. (I believe any performance difference of the generated code cannot be measured.)